wpull/urlprioritiser.py

```diff
@@ -0,0 +1,60 @@
+import copy
+from typing import Iterator, Tuple, Union
+from wpull.application.plugin import PluginFunctions, hook_interface
+from wpull.application.hook import HookableMixin, HookDisconnected
+from wpull.pipeline.item import URLRecord
+from wpull.url import URLInfo
+from wpull.urlfilter import BaseURLFilter
+
+
+class URLPrioritiser(HookableMixin):
+    '''Uses multiple URL filters to determine a URL's priority.'''
+    def __init__(self, priorities: Iterator[Tuple[BaseURLFilter, int]]):
+        super().__init__()
+        self._priorities = priorities
+        self.hook_dispatcher.register(PluginFunctions.get_priority)
+
+    @property
+    def priorities(self) -> Iterator[Tuple[BaseURLFilter, int]]:
+        return self._priorities
+
+    @staticmethod
+    @hook_interface(PluginFunctions.get_priority)
+    def plugin_get_priority(url_info: URLInfo, url_record: URLRecord) -> Union[int, None]:
+        '''Return the priority for this URL.
+
+        Args:
+            url_info (URLInfo): A representation of the URL in question
+            url_record (URLRecord): Information about the URL in the context of the crawl.
+                Note that the priority attribute will always be ``None`` because the hook is called *before* the internal rules.
+
+        Returns:
+            An integer to set the priority to that value, or None if the normal priorisation rules shall be consulted.
+        '''
+        return None
+
+    def _get_priority_from_hook(self, url_info: URLInfo, url_record: URLRecord) -> Union[int, None]:
+        '''Tries to call the get_priority hook. The hook can either return an int or None. If the hook is not connected, None is retured.'''
+        try:
+            # Only pass copies of the URLInfo and URLRecord instances to the hook to prevent modification.
+            # A deep copy is not actually needed as of writing this code because all attributes of both classes are immutable.
+            # However, to prevent future breakage if something changes in those classes, a deep copy is used anyway.
+            priority = self.hook_dispatcher.call(PluginFunctions.get_priority, copy.deepcopy(url_info), copy.deepcopy(url_record))
```

Comment:
deepcopy is almost always a bad idea because it defers copying responsibility to a slowish and general algorithm that may copy more than needed, or in some rare cases less than needed.   Something that needs to be copied should usually implement e.g. its own `copy` method that copies just what is necessary.  Since you say `deepcopy` isn't needed right now, maybe just don't do it?

Reply:
I agree. This currently also *always* `deepcopy`s the info and record objects, even if the hook isn't used at all.

There are probably three solutions to this:

1. Implement a `clone` method into `URLInfo` and `URLRecord` to create copies/clones.
2. Add a `freeze` method to these objects to make them immutable. I couldn't find any code that manipulates these objects after the initial creation. But this seems like a pretty good way to reintroduce the performance issues that apparently occurred with `FrozenDict` before its removal in 56138077.
3. Don't do anything about it in the code at all, just passing the original objects – maybe add a note in the documentation to not modify those from the hooks, then ignore it and let people shoot themselves in the foot if they really want to.

I'm tempted to go with number 3.

Reply:
3 sounds fine with me. I don't think we have enough hooks authors to justify guardrails here.